src/interpreter.rs

```diff
@@ -1924,27 +1924,26 @@ impl<'source> Interpreter<'source> {
                 head: rule_head,
                 bodies: rule_body,
             } => {
-                if matches!(rule_head, RuleHead::Func { .. }) {
-                    return Ok(());
-                }
+                if !matches!(rule_head, RuleHead::Func { .. }) {
```

Comment:
It might make sense for us to use something like [scopeguard](https://github.com/bluss/scopeguard) so that the cleanup code executes even in case of early returns.

Reply:
I will look at it.

Reply:
Unfortunately I didn't succeed to use scopeguard; by doing something like that:
```
let _guard = scopeguard::guard(rule, |rule| {
      self.processed.insert(rule);
    });
```

I had a borrowing error.
I can try a bit more later.